File: TwoSigma/longestChain.cpp
# include<iostream>
# include<unordered_map>
# include<vector>
# include<algorithm>
using namespace std;

int longestChain(vector<string>& words) {
  auto cmp = [](string& s1, string& s2){return (s1.size() < s2.size()) || (s1.size() == s2.size() && s1 < s2);};  
  sort(words.begin(), words.end(), cmp);
  unordered_map<string, int> dict;
  int ans = 0;
  for(auto word : words) {
    if(dict.count(word)) continue;
    dict[word] = 1;
    for(int i = 0; i < word.size(); ++i) {
      auto temp = word.substr(0, i) + word.substr(i + 1);
      if(dict.count(temp)) {
        dict[word] = max(dict[word], 1 + dict[temp]);
      }
    }
    ans = max(ans, dict[word]);
  }
  return ans;
}

int main() {
  vector<string> w1({"b", "bca", "ba", "bda", "bdca", "a"});
  vector<string> w2({"b"});
  vector<string> w3;
  vector<string> w4({"b","ba","bca","bdca","ebdca","b"});
  cout << longestChain(w1) << endl;
  cout << longestChain(w2) << endl;  
  cout << longestChain(w3) << endl;
  cout << longestChain(w4) << endl;  
  return 0;
}

File: TwoSigma/longestChain.py
def longestChain(words):
  """
  >>> longestChain(['b', 'bca', 'ba', 'bda', 'bdca', 'a'])
  4
  >>> longestChain(['b'])
  1
  >>> longestChain([])
  0
  >>> longestChain(["b","ba","bca","bdca","ebdca","b"])
  5
  """
  words.sort(cmp = lambda s1, s2: len(s1) < len(s2) || (len(s1) == len(s2) && s1 < s2))
  visit = {}
  ans = 0
  for word in words:
    if word in visit:
      continue
    visit[word] = 1
    for i in range(len(word)):
      temp = word[:i] + word[i + 1:]
      if temp in visit:
        visit[word] = max(visit[word], 1 + visit[temp])
    ans = max(ans, visit[word])
  return ans
        







